File: 496-next-greater-element-i/next-greater-element-i.java
class Solution {
    public int[] nextGreaterElement(int[] nums1, int[] nums2) {
        Map<Integer, Integer> nextGreater = new HashMap<>();
        Stack<Integer> stack = new Stack<>();
        
        // Traverse nums2 to find the next greater element for each element
        for (int num : nums2) {
            while (!stack.isEmpty() && stack.peek() < num) {
                nextGreater.put(stack.pop(), num);
            }
            stack.push(num);
        }
        
        // Initialize the result array with -1
        int[] result = new int[nums1.length];
        Arrays.fill(result, -1);
        
        // Fill the result array by looking up the next greater element for each element in nums1
        for (int i = 0; i < nums1.length; i++) {
            if (nextGreater.containsKey(nums1[i])) {
                result[i] = nextGreater.get(nums1[i]);
            }
        }
        
        return result;
    }
}

File: 496-next-greater-element-i/next-greater-element-i.cpp
class Solution {
public:
    vector<int> nextGreaterElement(vector<int>& nums1, vector<int>& nums2) {
        unordered_map<int, int> map;
        stack<int> st;
        for (int i = nums2.size() - 1; i >= 0; i--) {
            if (st.empty()) {
                map[nums2[i]] = -1;
            } else {
                while (!st.empty() && st.top() < nums2[i]) {
                    st.pop();
                }
                if (st.empty())
                    map[nums2[i]] = -1;
                else
                    map[nums2[i]] = st.top();
            }
            st.push(nums2[i]);
        }
        // for(auto i:map) cout<<i.first<<i.second;
        vector<int> ans;
        for (int i = 0; i < nums1.size(); i++) {
            ans.push_back(map[nums1[i]]);
        }
        return ans;
    }
};